compiler/rustc_trait_selection/src/traits/mod.rs

```diff
@@ -267,8 +267,29 @@ pub fn normalize_param_env_or_error<'tcx>(
     // parameter environments once for every fn as it goes,
     // and errors will get reported then; so outside of type inference we
     // can be sure that no errors should occur.
-    let mut predicates: Vec<_> =
-        util::elaborate(tcx, unnormalized_env.caller_bounds().into_iter()).collect();
+    let mut predicates: Vec<_> = util::elaborate(
+        tcx,
+        unnormalized_env.caller_bounds().into_iter().map(|predicate| {
+            if tcx.features().generic_const_exprs {
+                return predicate;
+            }
+
+            struct ConstNormalizer<'tcx>(TyCtxt<'tcx>);
+
+            impl<'tcx> TypeFolder<TyCtxt<'tcx>> for ConstNormalizer<'tcx> {
+                fn interner(&self) -> TyCtxt<'tcx> {
+                    self.0
+                }
+
+                fn fold_const(&mut self, c: ty::Const<'tcx>) -> ty::Const<'tcx> {
+                    c.eval(self.0, ty::ParamEnv::empty())
```

Comment:
should this perhaps be a `try_eval`? Otherwise maybe we want to pass in the unnormalized param-env?

Reply:
`ParamEnv::empty()` should be fine here :thinking: we check for `generic_const_exprs` and dont do anything if its enabled which means all anon consts should have no generics or param env _except_ for repeat expr counts, which can't appear here since we're folding `Ty`/`Const` not expressions.